### DISCUSSION TITLE
Race Conditions With Deregistering Self Hosted Runner and New Jobs Starting

### DISCUSSION THREAD

**Discussion by koreyGambill** [community/community#52867] [Actions]
### Select Topic Area

Product Feedback

### Body

Hello,

# Problem
There is a race condition between issuing a stop command and Github actions receiving and processing that command. Which is unavoidable due to network traffic delays, causing jobs to get cancelled.

My org is running self-hosted Github runners on ephemeral workloads. Before scaling down, we check that no jobs are running on the current node, then we run the actions-runner removal script, and then we terminate the node. Unfortunately, the actions-runner removal script takes 2 seconds, and in that time we often have a job placed on the node. We are having jobs fail every day because of this.

Stopping the runner with the following command takes almost 2 full seconds, subject to some variation
``` bash
if pgrep --full --ignore-case --list-full "runner.worker"; then
  echo "$(date +"%Y-%m-%d %H:%M:%S.%3N") - Runner is processing a job, don't remove"
else
  echo "$(date +"%Y-%m-%d %H:%M:%S.%3N") - removing runner from group"
  /opt/actions-runner/config.sh remove --token ${GH_TOKEN}
  echo "$(date +"%Y-%m-%d %H:%M:%S.%3N") - finished removing runner from group"
```
``` bash
# Output
2023-04-14 22:09:05.698 - removing runner from group
2023-04-14 22:09:07.642 - finished removing runner from group
```

# Proposed Solution
idea 1:
The API should expose a "Safe Stop" command which would guarantee that a worker was only stopped if it wasn't being picked to process a job.

idea 2:
Or the API could expose a "Pause Running New Jobs" command where the instance could be marked as "Paused" - giving four total states "Active", "Idle", "Offline", and finally "Paused". A paused runner would finish running it's current job, and move into a paused state. 

Either solution would allow us to locally script a safe way to kill nodes without a race condition.

# Other Considerations
We thought about using ephemeral workers, but some of our jobs spawn as many as 30 nodes to more quickly process tests. We are not aware of a way to see if a job was cancelled and tell those 30 nodes that they no longer need to exist. Furthermore, one of them could grab any of the other jobs in the queue and we could end up cancelling nodes that just received a job. So scaling without a pause or "stop unless running" feature is difficult.

# Workarounds
I don't currently know of a workaround, I would love some suggestions.
Our best idea currently is setting the ufw firewall on the host to block traffic from Github and hoping that new jobs fail to start on the runner so Github (hopefully) picks a new runner. This is a pretty hacky solution, but we may try this.

**Comment by koreyGambill** [comment]
Update: The ufw firewall did not work. Although the runner locally did not know that it had been assigned a job, the job was still assigned in github actions. So github actions does not seem to wait for confirmation from the runner that it is aware that it has been assigned a job.

I'd really like some suggestions. I really don't see how autoscaling self hosted runners is a possibility with github actions. Not without a bunch of randomly failing jobs, anyways. For a small team, the race condition is still happening many times per day.

**Reply by koreyGambill** [reply]
For anyone interested, my workaround was to write a lambda on a cron that monitors quickly failed jobs in github actions and issues a restart to the failed jobs within a workflow. It's not picky, it'll restart legitimate errors, but it works. It also won't restart a job more than twice (3rd run is the last run).

Disappointed I had to make this solution. We still run into failures daily because of this, but now they don't require manual intervention and they are more of an annoyance / small time sink for people waiting for builds.

**Comment by korran** [comment]
An idea I'm considering is to remove all custom labels from the runner, preventing new jobs from scheduling. Then I can wait awhile, and make sure there is no job running on the runner before I kill it.

**Comment by ChristopherHX** [comment]
Another idea, which worked in the past was

Create an ephemeral runner instance (auto remove after one job has run)

Calling this rest api https://docs.github.com/en/free-pro-team@latest/rest/actions/self-hosted-runners?apiVersion=2022-11-28#delete-a-self-hosted-runner-from-a-repository unconditonally with an authorized PAT / GitHub App token after 5min has passed, since the runner has been started. (The second step can be done from an autoscaler)

The important details here are that the rest api
- refuses to delete runners while they run jobs ( just a non success http error from delete self-hosted runner api)
- allows deleting runners while they are idle ( the runner notices it and exits with a message, no need to kill manually )

**Comment by adityamaru** [comment]
This is still quite painful to work around. We have added all the above protections to reduce the window of raciness but there are still situations in which de-registration of a runner succeeds and a few seconds later a job will be assigned to that runner. We rely on de-registration as our final signal to indicate that an ephemeral, self-hosted runner is safe to teardown.

@TingluoHuang I've noticed you're making significant improvements to the runner code in the latest release, and it is much appreciated. I was wondering if you could chime in here or route this to the right person? Is there a contract that ensures a runner is not assigned a job after that interface/method is called? Thank you!
